larpandoracontent/LArHelpers/LArClusterHelper.cc

```diff
@@ -600,6 +600,48 @@ void LArClusterHelper::GetCoordinateVector(const Cluster *const pCluster, Cartes
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+void LArClusterHelper::GetCaloHitList(const Cluster *const pCluster, CaloHitList &caloHitList)
+{
+    for (const OrderedCaloHitList::value_type &layerEntry : pCluster->GetOrderedCaloHitList())
+    {
+        for (const CaloHit *const pCaloHit : *layerEntry.second)
+            caloHitList.push_back(pCaloHit);
+    }
+
+    caloHitList.sort(LArClusterHelper::SortHitsByPosition);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArClusterHelper::GetCaloHitListInBoundingBox(const pandora::Cluster *const pCluster, const pandora::CartesianVector &lowerBound, 
+        const pandora::CartesianVector &upperBound , pandora::CaloHitList &caloHitList)
+{
+    for (const OrderedCaloHitList::value_type &layerEntry : pCluster->GetOrderedCaloHitList())
+    {
+        for (const CaloHit *const pCaloHit : *layerEntry.second)
+        {
+            const CartesianVector hitPosition = pCaloHit->GetPositionVector();
```

Comment:
Actually, the intention is not to gobble up all of the hits when the bounding box is 0, but that is what the code is doing; really nice catch.
The reason for the dimension checks is so that the bounding box still works in 2D; in the transverse alg, all of the cluster spans are constrained to the XZ plane.  Without the bounds check on Y, none of the contained hits would make it into the vector.

You are totally right though, if the box is 0 volume, all of the hits reach the box, which is not the intention.

This may be subjective, but I think that if the box's dimensions are all 0 then the calohit list should be empty.  If --any-- of the box dimensions are non-zero then the hits should be checked to see if they fit in those dimension(s).

What do you think @AndyChappell ?

Reply:
I think for the case of all dimensions being zero that's a more intuitive behaviour. Given the intention that the bounding box still works in 2D, I think checking the remaining non-zero dimensions makes sense - though does the 1D case make sense, or does it need to have two out of the three be non-zero?

Reply:
It's a good question. I guess there's a user case for getting all hits between 1D bounds, though that might be beyond the scope of this one function.

I did flick through the helper and could see a function that returns hits between 1D bounds.

Overall, I really don't mind.  I think there are upsides and downsides to having this function return bounded hits when only 1 dimension is used.

I personally don't have a use case for 1D bounding so I really don't have strong feelings about it.

Reply:
Similarly I don't have a strong view on which approach should be adopted, but it's probably worth documenting whichever behaviour you settle on in case it becomes significant in any future use case.

Reply:
I opted for allowing for the 1D case, partly because it made the logic easier 😅 
If all three dimensions are empty, the function now throws.

I hope this resolves this comment?

Reply:
Yep, that's fine with me.